zmq.cpp

```diff
@@ -68,26 +75,40 @@ void LoadBalancer::run(){
         }
         //handle clients request
         if (items[1].revents & ZMQ_POLLIN){
-            //first we get the client id
-            zmq::message_t msg_identity;
-            clients.recv(&msg_identity);
-            {
-                //then an empty frame
-                std::string empty = z_recv(clients);
-                assert(empty.size() == 0);
-            }
+            // The client request is a multi-part ZMQ message, we have to check every frame and be sure the multi-part message frame
+            // is composed as we wish, otherwise the multi-part message may be shifted unexpectedly.
+
+            // The multi-part ZMQ message should have 3 parts
+            // The first one is the ID of message
+            // The second one is an empty frame
+            // The third one is the real request
+            int more = 0;
+            size_t more_size = sizeof (more);
 
-            zmq::message_t msg_request;
-            clients.recv(&msg_request);
+            std::vector<zmq::message_t> frames{};
+            do {
+                zmq::message_t frame;
+                clients.recv(&frame);
+                frames.push_back(std::move(frame));
+                // Are there more frames coming?
+                clients.getsockopt(ZMQ_RCVMORE, &more, &more_size);
+            }while(more);
+
+            if (! is_valid_message(frames)) {
+                z_send(clients, "");
+                continue;
+            }
 
             std::string worker_addr = avalailable_worker.top();
             avalailable_worker.pop();
 
             z_send(workers, worker_addr, ZMQ_SNDMORE);
             z_send(workers, "", ZMQ_SNDMORE);
-            z_send(workers, msg_identity, ZMQ_SNDMORE);
+            // frames[0] is the id of message
+            z_send(workers, frames[0], ZMQ_SNDMORE);
             z_send(workers, "", ZMQ_SNDMORE);
-            z_send(workers, msg_request);
+            // frames[2] is the request
+            z_send(workers, frames[2]);
```

Comment:
Can't we just send the entire frames ? What would happened if the multi-part message is more than 3 frames in the future ?

Also, checking the returned value of a `z_send()` and logging erno values would be super nice.

Reply:
> Can't we just send the entire frame ?

no, when sending, you have to send frame by frame, just as receiving

> What would happened if the multi-part message is more than 3 frames in the future ?

by design, there must be 3 frames

> logging erno values would be super nice. 

let me try, the logging system is not at all the same in kraken and asgard...

Reply:
OK, logging in utils won't be easy ... it's out of the scope of this PR